mindmeld/models/taggers/pytorch_crf.py

```diff
@@ -0,0 +1,408 @@
+        elif self.optimizer_type == "adam":
+            self.optimizer = optim.Adam(self.parameters(), weight_decay=1e-5)
+
+        for epoch in range(self.epochs):
+            if _patience_counter >= self.patience:
+                break
+            self.train_one_epoch(train_dataloader)
+            dev_f1_score = self.run_predictions(dev_dataloader, calc_f1=True)
+            logger.debug("Epoch %s finished. Dev F1: %s", epoch, dev_f1_score)
+
+            if dev_f1_score <= best_dev_score:
+                _patience_counter += 1
+            else:
+                _patience_counter = 0
+                best_dev_score, best_dev_epoch = dev_f1_score, epoch
+                torch.save(self.state_dict(), self.best_model_save_path)
+                logger.debug("Model weights saved for best dev epoch %s.", best_dev_epoch)
+
+    def train_one_epoch(self, train_dataloader):
+        self.train()
+        train_loss = 0
+        for batch_idx, (inputs, mask, labels) in enumerate(train_dataloader):
+            self.optimizer.zero_grad()
+            loss = self.forward(inputs, labels, mask, drop_input=self.drop_input)
+            train_loss += loss.item()
+            loss.backward()
+            self.optimizer.step()
+            if batch_idx % 20 == 0:
+                logger.debug("Batch: %s Mean Loss: %s", batch_idx,
+                             (train_loss / (batch_idx + 1)))
+
+    def run_predictions(self, dataloader, calc_f1=False):
+        self.eval()
+        predictions = []
+        targets = []
+        with torch.no_grad():
+            for inputs, *mask_and_labels in dataloader:
+                if calc_f1:
+                    mask, labels = mask_and_labels
+                    targets.extend(torch.masked_select(labels, mask).tolist())
+                else:
+                    mask = mask_and_labels.pop()
+                preds = self.forward(inputs, None, mask)
+                predictions.extend([x for lst in preds for x in lst] if calc_f1 else preds)
+        if calc_f1:
+            dev_score = f1_score(targets, predictions, average='weighted')
+            return dev_score
+        else:
+            return predictions
+
+    def predict_marginals(self, X):
+        self.load_state_dict(torch.load(self.best_model_save_path))
+        inputs, seq_lens = self.encoder.get_tensor_data(X)
+        torch_dataset = TaggerDataset(inputs, seq_lens)
+
+        dataloader = DataLoader(torch_dataset, batch_size=512, shuffle=False, collate_fn=collate_tensors_and_masks)
+        marginals_dict = []
+        self.eval()
+        with torch.no_grad():
+            for inputs, mask in dataloader:
+                probs = self.compute_marginal_probabilities(inputs, mask).tolist()
+                mask = mask.tolist()
+                # If anyone has any suggestions on a cleaner way to do this, I am all ears!
+                marginals_dict.extend([[dict(zip(self.encoder.classes, token_probs)) \
```

Comment:
Can we break this nested list comprehension into individual structures? The three level zipping in a single statement might become tedious to follow later

Reply:
Refactored into a more readable format